### PR TITLE
DDS Security support

### PR DESCRIPTION
**Pull request by mikaelarguedas**
connects to ros2/ros2#339

### CODE REVIEW

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 1]
```diff
@@ -686,7 +688,10 @@ rmw_ret_t rmw_init()
   return RMW_RET_OK;
 }
 
-rmw_node_t * rmw_create_node(const char * name, const char * namespace_, size_t domain_id)
+rmw_node_t * create_node(
+  const char * name,
+  const char * namespace_,
+  ParticipantAttributes participantParam)
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: Variable naming.

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 2]
```diff
@@ -796,6 +794,81 @@ rmw_node_t * rmw_create_node(const char * name, const char * namespace_, size_t
...
+      property_policy.properties().emplace_back(
+        Property("dds.sec.auth.builtin.PKI-DH.private_key",
+        security_files_paths[2]));
+      property_policy.properties().emplace_back(
+        Property("dds.sec.crypto.plugin", "builtin.AES-GCM-GMAC"));
+      property_policy.properties().emplace_back(
+        Property("rtps.participant.rtps_protection_kind", "ENCRYPT"));
+      participantParam.rtps.properties = property_policy;
+    } else {
+      if (options->enforce_security) {
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: `elseif`?

**Comment by mikaelarguedas** [comment]
> What is the HAVE_SECURITY define needed for ? How is it being set?

This is defined by fast-RTPS in [config.h](https://github.com/ros2/rmw_fastrtps/pull/103/files#diff-8225b098f1ec4debcdcee03d65ae95d5R37) at compile time. We need to check it because people not building fastrtps with `SECURITY=ON` will not have the security headers and symbols available causing rmw_fastrtps_cpp to not compile.

File: rmw_fastrtps_cpp/src/functions.cpp (hunk 1 above)
Comment (mikaelarguedas) [inline rmw_fastrtps_cpp/src/functions.cpp]: good catch, just copy-n-paste the original variable without renaming it. Will fix

File: rmw_fastrtps_cpp/src/functions.cpp (hunk 2 above)
Comment (mikaelarguedas) [inline rmw_fastrtps_cpp/src/functions.cpp]: thanks, another simplification I forgot to refactor, will fix

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 3]
```diff
@@ -821,22 +826,15 @@ rmw_node_t * rmw_create_node(const char * name, const char * namespace_, size_t
     return NULL;
   }
 
-  eprosima::fastrtps::Log::SetVerbosity(eprosima::fastrtps::Log::Error);
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: Why has this been removed?

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 4]
```diff
@@ -919,6 +917,79 @@ rmw_node_t * rmw_create_node(const char * name, const char * namespace_, size_t
   return NULL;
 }
 
+bool rmw_get_security_file_paths(
+  std::array<std::string, 3> & security_files_paths, const char * node_secure_root)
+{
+  // here assume only 3 files for security
+  const char * file_names[3] = {"ca.cert.pem", "cert.pem", "key.pem"};
+  size_t num_files = 3;
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: Could this be `sizeof(...)` instead?

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 5]
```diff
@@ -919,6 +917,79 @@ rmw_node_t * rmw_create_node(const char * name, const char * namespace_, size_t
   return NULL;
 }
 
+bool rmw_get_security_file_paths(
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: This function is not part of the rmw interface and should therefore not have the `rmw_` prefix.

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 6]
```diff
@@ -919,6 +917,79 @@ rmw_node_t * rmw_create_node(const char * name, const char * namespace_, size_t
...
+  if (!name) {
+    RMW_SET_ERROR_MSG("name is null");
+    return NULL;
+  }
+
+  ParticipantAttributes participantAttrs;
+  participantAttrs.rtps.builtin.domainId = static_cast<uint32_t>(domain_id);
+  participantAttrs.rtps.setName(name);
+
+  if (security_options->security_root_path) {
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: You must check that `security_options` is not null before accessing its members.

Comment (mikaelarguedas) [inline rmw_fastrtps_cpp/src/functions.cpp]: done in https://github.com/ros2/rmw_fastrtps/pull/103/commits/4d4d8ff054e1d0c8036ea62ce34d016807f78d08

**Pull request merged**